### PR TITLE
[Lib] [CLI] Add support to handle multiple IO tools

### PR DESCRIPTION
**Pull request by pranavprakash20**
# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: cli/io/linux_untar.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+log = Log(__name__)
+
+
+class LinuxUntar:
+    def __init__(self, **config):
+        """
+        Args:
+            **config
+                        clients (list):     Client (s)
```
Comment (vamahaja) [inline cli/io/linux_untar.py:13]: Fix indentation

File: cli/io/linux_untar.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        Performs Linux untar on the given clients
+        """
+        threads = []
+        if not isinstance(self.clients, list):
+            self.clients = [self.clients]
+
+        for client in self.clients:
+            # Download linux untar to root
+            cmd = (
```
Comment (vamahaja) [inline cli/io/linux_untar.py:33]: Make it parameterize

File: cli/io/linux_untar.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                "wget https://cdn.kernel.org/pub/linux/kernel/"
+                "v5.x/linux-5.4.54.tar.xz"
+            )
+            client.exec_command(cmd=cmd, sudo=True)
+
+            for directory in self.dirs:
+                # copy linux tar to dir
+                cmd = "cp /root/linux-5.4.54.tar.xz {}/{}".format(
```
Comment (vamahaja) [inline cli/io/linux_untar.py:41]: same as above

File: cli/io/linux_untar.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                    self.mountpoint, directory
+                )
+                client.exec_command(cmd=cmd, sudo=True)
+
+                # Start linux untar
+                cmd = "cd {}/{};tar -xvf linux-5.4.54.tar.xz".format(
```
Comment (vamahaja) [inline cli/io/linux_untar.py:47]: same as above

File: cli/io/linux_untar.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                        mountpoint(str):    Mount point where the volume is
+                                            mounted.
+                        dirs(tuple):        A tuple of dirs where untar has to
+                                            started. (Default:('.'))
+        """
+        self.clients = config["clients"]
+        self.mountpoint = config["clients"]
+        self.dirs = config.get("dirs", ("."))
+
+    def linux_untar(self):
```
Comment (vamahaja) [inline cli/io/linux_untar.py:23]: change this to some common name like `start`

File: cli/io/linux_untar.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        Args:
+            **config
+                        clients (list):     Client (s)
+                        mountpoint(str):    Mount point where the volume is
+                                            mounted.
+                        dirs(tuple):        A tuple of dirs where untar has to
+                                            started. (Default:('.'))
+        """
+        self.clients = config["clients"]
+        self.mountpoint = config["clients"]
```
Comment (vamahaja) [inline cli/io/linux_untar.py:20]: `config["clients"]` ???

File: cli/io/__init__.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+log = Log(__name__)
+
+
+class IO:
+    """Interface for common cloud operations"""
+
+    def __init__(self, io_type, **config):
+        log.info(f"Initiating instance for IO '{io_type}'")
+        available_ios = ["specstorage", "linux_untar", "fio"]
```
Comment (vamahaja) [inline cli/io/__init__.py:15]: supported platforms

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Comment by pranavprakash20** [comment]
Added support for smallfile and FIO

**Comment by pranavprakash20** [comment]
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-QNE0J1/

**Review by MohitBis — DISMISSED** [review]
LGTM

File: cli/io/fio.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
+import itertools
+
+from cli import Cli
+from cli.utilities.packages import Package
+
+
+class FioError(Exception):
```
Comment (vamahaja) [inline cli/io/fio.py:7]: Change this exception to `IOEroro` and move to - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py

File: cli/io/fio.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    pass
+
+
+class Fio(Cli):
+    def __init__(self, client):
+        super(Fio, self).__init__(client)
+        self.client = client
+
+    def install_fio(self):
```
Comment (vamahaja) [inline cli/io/fio.py:16]: This should be part of installation step. Here in library assume that this package is present.

File: cli/io/fio.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        except Exception:
+            raise FioError("Fio installation failed")
+
+    def create_config(
+        self, ioengine, mount_dir, workloads, sizes, iodepth_values, numjobs
+    ):
+        """
+        Generates all the combimations of workloads sizes, iodepth as per the fio_config
+        Sample fio_config:
+        fio_config:
```
Comment (vamahaja) [inline cli/io/fio.py:32]: I assume these are test suite configs. Remove this and add description for parameters passed to this method

File: cli/io/fio.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+          workload_params:
+            random_rw:
+              rw: ["read","write"]
+              rwmixread: "70"
+              bs: 8k
+              numjobs: "4"
+              iodepth: ["4","8","16","32"]
+        """
+
+        fio_configs = []
```
Comment (vamahaja) [inline cli/io/fio.py:48]: no need to use fio as prefix or postfix in this libarary. Change this in all places.

File: cli/io/fio.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            print(config)
+            remote_file.write(config)
+            remote_file.flush()
+            fio_filenames.append(filename)
+        return fio_filenames
+
+    def _generate_fio_config(
+        self, ioengine, mount_dir, workload, size, iodepth, numjobs
+    ):
+        fio_config = f"""
```
Comment (vamahaja) [inline cli/io/fio.py:74]: Is this correct indentation? no pep8 error ?

File: cli/io/fio.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+            )
+            print(config)
+            remote_file.write(config)
+            remote_file.flush()
+            fio_filenames.append(filename)
+        return fio_filenames
+
+    def _generate_fio_config(
+        self, ioengine, mount_dir, workload, size, iodepth, numjobs
+    ):
```
Comment (vamahaja) [inline cli/io/fio.py:73]: Add description to method

File: cli/io/fio.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+numjobs={int(numjobs)}
+iodepth= {iodepth}
+"""
+        return fio_config
+
+    def run(self, mount_dir, ioengine, workloads, sizes, iodepth_values, numjobs):
+        """
+        Run Fio
+        Args:
+            mount_dir (str): Mounting directory
```
Comment (vamahaja) [inline cli/io/fio.py:95]: missing description for some parameters

File: cli/io/small_file.py [hunk 15]
```diff
@@ -0,0 +1,77 @@
+from random import choice
+from string import ascii_lowercase, digits
+
+from cli import Cli
+
+
+class SmallfileError(Exception):
```
Comment (vamahaja) [inline cli/io/small_file.py:7]: Same as above

File: cli/io/small_file.py [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    pass
+
+
+class SmallFile(Cli):
+    def __init__(self, client):
+        super(SmallFile, self).__init__(client)
+        self.client = client
+
+    def pull_smallfile_repo(self):
```
Comment (vamahaja) [inline cli/io/small_file.py:16]: Same as above

File: cli/io/small_file.py [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+        operations,
+        threads=10,
+        file_size=4,
+        files=1000,
+        files_per_dir=10,
+        dirs_per_dir=2,
+        record_size=128,
+    ):
+        """
+        Runs small file IO
```
Comment (vamahaja) [inline cli/io/small_file.py:41]: This is correct, follow same format for Fio lib

File: cli/io/small_file.py [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        if not isinstance(operations, list):
+            operations = [operations]
+
+        # Install Fio
+        self.pull_smallfile_repo()
+
+        dir = "".join(choice(ascii_lowercase + digits) for _ in range(10))
+
+        # Create tmp folder
+        cmd = "mkdir -p /var/tmp/smf"
```
Comment (vamahaja) [inline cli/io/small_file.py:61]: don't we have method for this one in `CephNode` object ?

File: cli/io/small_file.py [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+        self.execute(sudo=True, cmd=cmd)
+
+        # Create dir inside mount dir for IO
+        cmd = f"mkdir -p {mount_dir}/{dir}"
+        self.execute(sudo=True, cmd=cmd)
+
+        # Run smallfile
+        for operation in operations:
+            cmd = (
+                f"python3 smallfile/smallfile_cli.py --operation {operation} --threads {threads} "
```
Comment (vamahaja) [inline cli/io/small_file.py:71]: `python3 smallfile/smallfile_cli.py` create global variable for this script path and make it configuarable

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cli/io/small_file.py [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+            files (str): No of files
+            files_per_dir (str):  No files per directory created
+            dirs_per_dir (str): No of dirs per directory created
+            record_size (str): Record size
+        """
+        if not isinstance(operations, list):
+            operations = [operations]
+
+        # Install Fio
+        self.pull_smallfile_repo()
```
Comment (vamahaja) [inline cli/io/small_file.py:56]: Move these setup part in cli/utilities/utils.py and use names like -

- setup_smallfile_io
- setup_fio

**Comment by pranavprakash20** [comment]
@vamahaja All review comments addressed, please re-review

File: cli/utilities/utils.py [hunk 21]
```diff
@@ -704,3 +705,30 @@ def get_file_owner(filepath, clients):
             f"The file '{filepath}' is created by an unknown user with UID: {uid} and GID: {gid}."
         )
         return None
+
+
+def install_fio(client):
```
Comment (pranavprakash20) [inline cli/utilities/utils.py:710]: <comment> chage to setup_fio

**Review by vamahaja — APPROVED** [review]

**Pull request merged**